qiita_ware/search.py

```diff
@@ -0,0 +1,101 @@
+from collections import Counter, defaultdict
+
+from future.utils import viewvalues, viewitems
+
+from qiita_db.study import Study
+from qiita_db.data import ProcessedData
+
+
+def count_metadata(results, meta_cols):
+    """Counts the metadata found in a search, and returns these counts
+
+    Parameters
+    ----------
+    results : dict of lists of list
+        results in the format returned by the qiita_db search obj
+    meta_cols : list
+        metadata column names searched for, as returned by qiita_db search obj
+
+    Returns
+    -------
+    fullcount : dict of dicts
+        counts for each found metadata value over all studies, in the format
+        {meta_col1: {value1: count, value2: count, ...}, ...}
+    studycount : dict of dict of dicts
+        counts for each found metadata value for each study, in the format
+        {study_id: {meta_col1: {value1: count, value2: count, ...}, ...}, ...}
+    """
+    def double_comprehension(results):
+        for samples in viewvalues(results):
+            for sample in samples:
+                yield sample
+
+    fullcount = {}
+    # rearrange all samples so that each metadata column found is its own list
+    meta_vals = zip(*double_comprehension(results))
+    for pos, cat in enumerate(meta_cols):
+        # use Counter object to count all metadata values for a column
+        # pos+1 so we skip the sample names list
+        fullcount[cat] = Counter(meta_vals[pos + 1])
+
+    # Now get metadata counts for each study, removing sample ids as before
+    studycount = {}
+    for study_id in results:
+        hold = {}
+        # zip all samples for a given study so that each metadata column found
+        # is its own list
+        meta_vals = zip(*(sample for sample in results[study_id]))
+        for pos, cat in enumerate(meta_cols):
+            # use Counter object to count all metadata values for a column
+            # pos+1 so we skip the sample names list
+            hold[cat] = Counter(meta_vals[pos + 1])
+        studycount[study_id] = hold
+
+    return fullcount, studycount
+
+
+def filter_by_processed_data(results, datatypes=None):
+    """Filters results to what is available in each processed data
+
+    Parameters
+    ----------
+    results : dict of lists of list
+        results in the format returned by the qiita_db search obj
+    datatypes : list of str, optional
+        Datatypes to selectively return. Default all datatypes available
+
+    Returns
+    -------
+    study_proc_ids : dict of dicts of lists
+        Processed data ids with samples for each study, in the format
+        {study_id: {datatype: [proc_id, proc_id, ...], ...}, ...}
+    proc_data_samples : dict of lists
+        Samples available in each processed data id, in the format
+        {proc_data_id: [samp_id1, samp_id2, ...], ...}
+    samples_meta : dict of lists
```

Comment:
from the code below, the sample IDs contained in here will be a superset of the sample IDs described in `proc_data_samples`. Are those objects expected to be 1-1? If not, it would be useful to include a mention of this in the docstring